crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -323,6 +336,139 @@ impl Timeline {
         }
     }
 
+    /// Toggle a reaction on an event
+    ///
+    /// Adds or redacts a reaction based on the state of the reaction at the
+    /// time it is called.
+    ///
+    /// When redacting an event, the redaction reason is not sent.
+    ///
+    /// Ensures that only one reaction is sent at a time to avoid race
+    /// conditions and spamming the homeserver with requests.
+    ///
+    /// FIXME: If called multiple times in quick succession on the same
+    /// reaction, the function will drop requests until the previous request
+    /// is complete.
+    pub async fn toggle_reaction(&self, annotation: &Annotation) -> Result<(), Error> {
+        let txn_id = TransactionId::new();
+        let user_id = self.user_id()?;
+
+        // Check if the reaction already exists
+        let find_related_event = || async {
+            self.item_by_event_id(&annotation.event_id).await.ok_or(Error::FailedTogglingReaction)
+        };
+
+        let related_event = find_related_event().await?;
+
+        let user_reactions =
+            related_event.reactions().get(&annotation.key).map(|group| group.by_sender(&user_id));
+
+        // Find an existing reaction to redact, if any
+        let to_redact = user_reactions
+            .map(|mut reactions| reactions.find_map(|(_txid, event_id)| event_id))
+            .unwrap_or(None);
+
+        // Wait for any ongoing reaction requests to finish
+        let _lock = self.reaction_lock.lock().await;
```

Comment:
Instead of introducing a new lock with no associated data, can you move most of the logic of this method into a separate method on `TimelineInner` that locks the required state once?

Reply:
I am not sure - just to explain this a bit more: `reaction_lock` is being used to create a critical section below, creating a queue for threads to enter it. So it is not actually locking access to one particular bit of state but it's affecting the control flow, ensuring that threads don't perform the homeserver request and reconcile the responses concurrently for the same reaction.

Is that still something we'd want to move inside `TimelineInner`? If so do you have an idea of how it might be done?

Reply:
Well, what I'm getting at is that the inner timeline state is being locked many more times than it should. It should really just be locked once before the request, then once after. Otherwise it's way too easy to race with a different concurrent task that updates another part of the same item or whatever.

I'm also not convinced about having an extra critical section at this level. Why would this function ever be called twice in parallel?

Reply:
Perhaps I can describe the problem I'm trying to solve a bit better:

We want to be able to toggle the local echo of a reaction without 1) always triggering a request to the homeserver 2) being blocked by an ongoing request to the homeserver.

I understood `toggle_reaction` could be called twice in parallel if a user quickly adds and redacts the same reaction. If we were only to lock the timeline state once before the request, then once after, then a user could quickly and repeatedly toggle a reaction and trigger many concurrent requests to the homeserver. The problem then is that the winning response would be pretty random.

Reply:
But with what you've implemented, we still send N requests when the user presses the reaction button N times, right?

Reply:
Not necessarily - we check the state to determine if the request still needs to be done ([here](https://github.com/matrix-org/matrix-rust-sdk/pull/2118/files#diff-7f49982c21dd9af0fcce15d1d1349716f42c41268483df13c1dc8086c6448d57R389-R392) or [here](https://github.com/matrix-org/matrix-rust-sdk/pull/2118/files#diff-7f49982c21dd9af0fcce15d1d1349716f42c41268483df13c1dc8086c6448d57R381-R384))

Reply:
My understanding is that it will only cancel operations that would have done the same thing, for instance there's no reaction and you rapidly emit 10 requests to toggle, then in that case only it will abort the 9 following toggles (since they were attempting to do the same thing).

Agreed that the state lock is being locked more than it should be (it's implicitly happening in `item_by_event_id`, and before making the local echo), so that would justify moving this over to `TimelineInner`.

To achieve the same as your lock, maybe there could, in the state guarded by the existing lock, a set of in-flight requests, keyed by event id and target reaction. That would be filled before sending the request, and we'd remove the entry after processing the response (independently of the response's result). When starting the request, after taking the lock we can look into the set to see if there's already a pending request for the same action. (That sounds like a mechanism that could be useful for other request/response mechanisms, too, so maybe there's already something like that?)

That being said, it's not clear to me that this isn't a responsibility for the caller: we can't really infer how they want to manage consecutive quick reaction toggles. For instance, if as a user I'm pressing by mistake a button to enable a reaction, I'm likely to press it again to cancel quickly, quickly enough that the first press hasn't been ack'd by the server; and the system as is right now would simply abort my second request, not doing what I intended.

Reply:
Thanks for all your feedback.

I hope my latest changes should improve on some of the issues raised:
- The behaviour improved and reactions are no longer dropped (see the integration test which shows this in action)
- The timeline is now locked only once before a network request and once after to reduce the risk of race conditions
- There is no longer any unnecessary critical section in `Timeline` - the new mechanism uses `TimelineInner` to do the locking